### ISSUE TITLE
Batch norm optimizations

### ISSUE THREAD

**Issue by timmoon10** [LLNL/lbann#618]
I have two unrelated but overlapping optimizations for batch norm.

### Fused scale-bias weights
In Resnet-50, approximately 2/3 of the backprop allreduces originate from batch norm. Merging the scale and bias weights in batch norm would thus reduce the number of backprop allreduces by 1/3 and move each allreduce very slightly away from the latency-dominated regime. The resulting fused weights are somewhat less interpretable and more cumbersome, but the performance boost is probably worth it. This is an easy change to make.

### Separating the affine transformation into a distinct layer
The original batch norm paper includes a normalization step (subtract mean and divide by variance) and an affine step (apply scale and bias). These are logically discrete steps and many models apply the normalization step without the affine step (for instance, [Inception v3](https://github.com/keras-team/keras-applications/blob/master/keras_applications/inception_v3.py) applies a bias without scaling). TensorFlow, PyTorch, and Caffe make the affine step optional and Caffe removes it entirely. We currently support this if the user manually sets the batch norm weights with a scale of one, bias of zero, and no optimizer, but this is very unwieldy and we still do some unnecessary computation. I propose we follow Caffe's approach and implement a separate affine layer (with fused scale and bias weights). This will interact very naturally with our Keras frontend (the affine step option will just insert an affine layer after the batch norm), but will requiring modifying all prototext models that contain batch norm.

Pinging @ndryden and @JaeseungYeom.

**Comment by ndryden** [comment]
Another optimization we should consider: Since we have custom implementations of both batchnorm and ReLU, we should support fusing them into a single kernel.

**Comment by ndryden** [comment]
After recent discussion, there's a performance need for both fusing the weights and to fuse the allreduces done in intra-node/global batchnorm.
